meteor/diffmaps.py

```diff
@@ -0,0 +1,195 @@
+    return amplitudes.to_numpy() * np.exp(1j * np.deg2rad(phases.to_numpy()))
+
+
+def _complex_array_to_rs_dataseries(
+    complex_array: np.ndarray,
+    index: pd.Index,
+) -> tuple[rs.DataSeries, rs.DataSeries]:
+    # Extract amplitude from the complex array (magnitude)
+    amplitudes = rs.DataSeries(np.abs(complex_array), index=index)
+    amplitudes = amplitudes.astype(rs.StructureFactorAmplitudeDtype())  # Convert dtype
+
+    # Extract phase from the complex array (angle, in degrees)
+    phases = rs.DataSeries(np.rad2deg(np.angle(complex_array)), index=index)
+    phases = phases.astype(rs.PhaseDtype())  # Convert dtype
+
+    return amplitudes, phases
+
+
+def compute_deltafofo(
+    dataset: rs.DataSet,
+    *,
+    native_amplitudes: str,
+    derivative_amplitudes: str,
+    native_phases: str,
+    derivative_phases: str | None = None,
+    inplace: bool = False,
+) -> rs.DataSet | None:
+    """
+    Compute amplitude and phase differences between native and derivative datasets.
+    """
+
+    if not inplace:
+        dataset = dataset.copy()
+
+    # Convert native and derivative amplitude/phase pairs to complex arrays
+    native_complex = _rs_dataseries_to_complex_array(
+        dataset[native_amplitudes], dataset[native_phases]
+    )
+
+    if derivative_phases is not None:
+        derivative_complex = _rs_dataseries_to_complex_array(
+            dataset[derivative_amplitudes], dataset[derivative_phases]
+        )
+    else:
+        # If no derivative phases are provided, assume they are the same as native phases
+        derivative_complex = _rs_dataseries_to_complex_array(
+            dataset[derivative_amplitudes], dataset[native_phases]
+        )
+
+    # Compute complex differences
+    delta_complex = derivative_complex - native_complex
+
+    # Convert back to amplitude and phase DataSeries
+    delta_amplitudes, delta_phases = _complex_array_to_rs_dataseries(
+        delta_complex, dataset.index
+    )
+
+    # Add results to dataset
+    dataset["DeltaFoFo"] = delta_amplitudes
+    dataset["DeltaPhases"] = delta_phases
+
+    # Reset to Phase dtype
+    dataset["DeltaPhases"] = dataset["DeltaPhases"].astype("Phase")
+
+    if inplace:
+        return None
+    else:
+        return dataset
+
+
+def compute_kweights(
+    df: rs.DataSeries, sigdf: rs.DataSeries, kweight: float
+) -> rs.DataSeries:
+    """
+    Compute weights for each structure factor based on DeltaF and its uncertainty.
+    """
+    w = 1 + (sigdf**2 / (sigdf**2).mean()) + kweight * (df**2 / (df**2).mean())
+    return w**-1
+
+
+def compute_kweighted_deltafofo(
+    dataset: rs.DataSet,
+    *,
+    native_amplitudes: str,
+    derivative_amplitudes: str,
+    native_phases: str,
+    derivative_phases: str | None = None,
+    sigf_native: str,
+    sigf_deriv: str,
+    kweight: float = 1.0,
+    optimize_kweight: bool = False,
+    inplace: bool = False,
+) -> rs.DataSet | None:
```

Comment:
same comment about in place here

Reply:
I like inplace: otherwise we either (1) copy everytime, not always efficient/necessary or (2)need to make a new dataset with spacegroup cell etc